src/nvrpc/BidirectionalStreamingLifeCycle.h

```diff
@@ -0,0 +1,366 @@
+{
+  return (this->*m_NextState)(ok);
+}
+
+
+template <class Request, class Response>
+BidirectionalStreamingLifeCycle<
+    Request, Response>::BidirectionalStreamingLifeCycle()
+    : m_ReadStateContext(static_cast<IContext*>(this)),
+      m_WriteStateContext(static_cast<IContext*>(this))
+{
+  m_ReadStateContext.m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateRequestDone;
+  m_WriteStateContext.m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateResponseDone;
+}
+
+template <class Request, class Response>
+void
+BidirectionalStreamingLifeCycle<Request, Response>::Reset()
+{
+  std::queue<RequestType> empty_request_queue;
+  std::queue<ResponseType> empty_response_queue;
+  std::queue<ResponseType> empty_response_write_back_queue;
+  OnLifeCycleReset();
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
+    m_Executing = false;
+    m_RequestQueue.swap(empty_request_queue);
+    m_ResponseQueue.swap(empty_response_queue);
+    m_ResponseWriteBackQueue.swap(empty_response_write_back_queue);
+    m_Context.reset(new ::grpc::ServerContext);
+    m_ReaderWriter.reset(
+        new ::grpc::ServerAsyncReaderWriter<ResponseType, RequestType>(
+            m_Context.get()));
+  }
+  m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateInitializedDone;
+  m_QueuingFunc(m_Context.get(), m_ReaderWriter.get(), IContext::Tag());
+}
+
+// The following are a set of functions used as function pointers
+// to keep track of the state of the context.
+template <class Request, class Response>
+bool
+BidirectionalStreamingLifeCycle<Request, Response>::StateInitializedDone(
+    bool ok)
+{
+  if (!ok)
+    return false;
+
+  OnLifeCycleStart();
+  // Start reading once connection is created
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
+    m_RequestQueue.emplace();
+  }
+  m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateRequestDone;
+  m_ReaderWriter->Read(
+      &m_RequestQueue.back(), m_ReadStateContext.IContext::Tag());
+  return true;
+}
+
+// If the Context.m_NextState is at this state or at StateResponseDone state,
+// it will keep reading requests from the stream until no more requests will
+// be read (Read() brings back status ok==false)
+template <class Request, class Response>
+bool
+BidirectionalStreamingLifeCycle<Request, Response>::StateRequestDone(bool ok)
+{
+  // No more message to be read from this stream, however, if it is executing
+  // a request, then a ServerReaderWriter::Write() will be called. In that case,
+  // let WriteStateContext handle the reset procedure.
+  if (!ok) {
+    {
+      std::unique_lock<std::mutex> lock(m_QueueMutex);
+      if (m_Executing)
+        return true;
+    }
+    return false;
+  }
+
+  // Successfully receive request
+  bool should_execute = false;
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
```

Comment:
no condition variable, so we should probably be using `std::lock_guard` until we upgraded C++17 and then we can use `std::scoped_lock`